examples/generative/gan_ada.py

```diff
@@ -0,0 +1,729 @@
+
+The authors of [StyleGAN2-ADA](https://arxiv.org/abs/2006.06676) show that discriminator
+overfitting can be an issue in GANs, especially when only low amounts of training data is
+available. They propose Adaptive Discriminator Augmentation to mitigate this issue.
+
+Applying data augmentation to GANs however is not straightforward. Since the generator is
+updated using the discriminator's gradients, if the generated images are augmented, the
+augmentation pipeline has to be differentiable and also has to be GPU-compatible for
+computational efficiency. Luckily, the [Keras image augmentation
+layers](https://keras.io/api/layers/preprocessing_layers/image_augmentation/) fulfill
+both these requirements, and are therefore very well suited for this task.
+
+### Invertible data augmentation
+
+A possible difficulty when using data augmentation in generative models is the issue of
+["leaky augmentations" (section 2.2)](https://arxiv.org/abs/2006.06676), namely when the
+model generates images that are already augmented. This would mean that it was not able
+to separate the augmentation from the underlying data distribution, which can be caused
+by using non-invertible data transformations. For example, if either 0, 90, 180 or 270
+degree rotations are performed with equal probability, the original orientation of the
+images is impossible to infer, and this information is destroyed.
+
+A simple trick to make data augmentations invertible is to only apply them with some
+probability. That way the original version of the images will be more common, and the
+data distribution can be infered. By properly choosing this probability, one can
+effectively regularize the discriminator without making the augmentations leaky.
+
+"""
+
+"""
+## Setup
+"""
+
+import matplotlib.pyplot as plt
+import tensorflow as tf
+import tensorflow_datasets as tfds
+
+from tensorflow import keras
+from tensorflow.keras import layers
+
+"""
+## Hyperparameterers
+"""
+
+# data
+num_epochs = 10  # train for 400 epochs for good results
+image_size = 64
+kid_image_size = 75  # resolution of KID measurement
+padding = 0.25
+dataset_name = "caltech_birds2011"
+
+# adaptive discriminator augmentation
+max_translation = 0.125
+max_rotation = 0.125
+max_zoom = 0.25
+target_accuracy = 0.85
+integration_steps = 1000
+
+# architecture
+noise_size = 64
+depth = 4
+width = 128
+leaky_relu_slope = 0.2
+dropout_rate = 0.4
+
+# optimization
+batch_size = 128
+learning_rate = 2e-4
+beta_1 = 0.5  # not using the default value of 0.9 is important
+ema = 0.99
+
+"""
+## Data pipeline
+
+In this example, we will use the
+[Caltech Birds (2011)](https://www.tensorflow.org/datasets/catalog/caltech_birds2011) dataset for
+generating images of birds, which is a diverse natural dataset containing less then 6000
+images for training. When working with such low amounts of data, one has to take extra
+care to retain as high data quality as possible. In this example, I use the provided
```

Comment:
Generally speaking, prefer using the academic "we" instead of "I". It also helps the reader feel more involved!